File: sdks/local-rp/java/src/test/java/community/catalyst/linkkeys/localrp/testutil/MiniJson.java
package community.catalyst.linkkeys.localrp.testutil;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * A minimal, dependency-free JSON parser for reading the conformance vector
 * fixtures in tests. This SDK has zero runtime dependencies (design doc:
 * "Zero external dependencies (test framework excepted)"), and a JSON
 * library counts as a dependency even at test scope, so the conformance
 * vector loader hand-parses JSON the same way this SDK hand-parses CBOR.
 *
 * <p>Only what {@code sdks/local-rp/conformance/*.json} needs: objects,
 * arrays, strings (with standard escapes), numbers, booleans, and null.
 * Numbers are returned as {@link Double} or {@link Long} depending on
 * whether they contain a {@code .}/exponent; the fixtures only ever need
 * integers, so callers use {@link JsonValue#asLong()}.
 */
public final class MiniJson {
    private MiniJson() {}

    public static JsonValue parse(String text) {
        Parser p = new Parser(text);
        p.skipWhitespace();
        JsonValue v = p.parseValue();
        p.skipWhitespace();
        if (!p.atEnd()) {
            throw new IllegalArgumentException("trailing content after JSON value at position " + p.pos);
        }
        return v;
    }

    /** A parsed JSON value: object ({@code Map<String,JsonValue>}), array, string, number, boolean, or null. */
    public static final class JsonValue {
        private final Object raw;

        private JsonValue(Object raw) {
            this.raw = raw;
        }

        public boolean isNull() {
            return raw == null;
        }

        public String asString() {
            if (raw instanceof String s) {
                return s;
            }
            throw new IllegalStateException("not a string: " + raw);
        }

        public boolean asBoolean() {
            if (raw instanceof Boolean b) {
                return b;
            }
            throw new IllegalStateException("not a boolean: " + raw);
        }

        public long asLong() {
            if (raw instanceof Long l) {
                return l;
            }
            if (raw instanceof Double d) {
                return d.longValue();
            }
            throw new IllegalStateException("not a number: " + raw);
        }

        @SuppressWarnings("unchecked")
        public List<JsonValue> asArray() {
            if (raw instanceof List<?> list) {
                return (List<JsonValue>) list;
            }
            throw new IllegalStateException("not an array: " + raw);
        }

        @SuppressWarnings("unchecked")
        public Map<String, JsonValue> asObject() {
            if (raw instanceof Map<?, ?> map) {
                return (Map<String, JsonValue>) map;
            }
            throw new IllegalStateException("not an object: " + raw);
        }

        public JsonValue get(String key) {
            JsonValue v = asObject().get(key);
            if (v == null) {
                throw new IllegalStateException("missing key '" + key + "' in " + raw);
            }
            return v;
        }

        public JsonValue getOrNull(String key) {
            return asObject().get(key);
        }

        /** {@code get(key).asString()}, or {@code null} if the key is absent or its value is JSON null. */
        public String getStringOrNull(String key) {
            JsonValue v = getOrNull(key);
            return (v == null || v.isNull()) ? null : v.asString();
        }
    }

    private static final class Parser {
        private final String s;
        private int pos = 0;

        Parser(String s) {
            this.s = s;
        }

        boolean atEnd() {
            return pos >= s.length();
        }

        void skipWhitespace() {
            while (pos < s.length() && Character.isWhitespace(s.charAt(pos))) {
                pos++;
            }
        }

        private char peek() {
            if (pos >= s.length()) {
                throw new IllegalArgumentException("unexpected end of JSON input");
            }
            return s.charAt(pos);
        }

        private void expect(char c) {
            if (peek() != c) {
                throw new IllegalArgumentException("expected '" + c + "' at position " + pos + ", got '" + peek() + "'");
            }
            pos++;
        }

        JsonValue parseValue() {
            skipWhitespace();
            char c = peek();
            return switch (c) {
                case '{' -> parseObject();
                case '[' -> parseArray();
                case '"' -> new JsonValue(parseString());
                case 't' -> {
                    expectLiteral("true");
                    yield new JsonValue(Boolean.TRUE);
                }
                case 'f' -> {
                    expectLiteral("false");
                    yield new JsonValue(Boolean.FALSE);
                }
                case 'n' -> {
                    expectLiteral("null");
                    yield new JsonValue(null);
                }
                default -> parseNumber();
            };
        }

        private void expectLiteral(String literal) {
            if (!s.regionMatches(pos, literal, 0, literal.length())) {
                throw new IllegalArgumentException("expected '" + literal + "' at position " + pos);
            }
            pos += literal.length();
        }

        private JsonValue parseObject() {
            expect('{');
            Map<String, JsonValue> map = new LinkedHashMap<>();
            skipWhitespace();
            if (peek() == '}') {
                pos++;
                return new JsonValue(map);
            }
            while (true) {
                skipWhitespace();
                String key = parseString();
                skipWhitespace();
                expect(':');
                JsonValue value = parseValue();
                map.put(key, value);
                skipWhitespace();
                char c = peek();
                if (c == ',') {
                    pos++;
                    continue;
                }
                if (c == '}') {
                    pos++;
                    break;
                }
                throw new IllegalArgumentException("expected ',' or '}' at position " + pos);
            }
            return new JsonValue(map);
        }

        private JsonValue parseArray() {
            expect('[');
            List<JsonValue> list = new ArrayList<>();
            skipWhitespace();
            if (peek() == ']') {
                pos++;
                return new JsonValue(list);
            }
            while (true) {
                list.add(parseValue());
                skipWhitespace();
                char c = peek();
                if (c == ',') {
                    pos++;
                    continue;
                }
                if (c == ']') {
                    pos++;
                    break;
                }
                throw new IllegalArgumentException("expected ',' or ']' at position " + pos);
            }
            return new JsonValue(list);
        }

        private String parseString() {
            expect('"');
            StringBuilder sb = new StringBuilder();
            while (true) {
                char c = s.charAt(pos++);
                if (c == '"') {
                    break;
                }
                if (c == '\\') {
                    char esc = s.charAt(pos++);
                    switch (esc) {
                        case '"' -> sb.append('"');
                        case '\\' -> sb.append('\\');
                        case '/' -> sb.append('/');
                        case 'b' -> sb.append('\b');
                        case 'f' -> sb.append('\f');
                        case 'n' -> sb.append('\n');
                        case 'r' -> sb.append('\r');
                        case 't' -> sb.append('\t');
                        case 'u' -> {
                            String hex = s.substring(pos, pos + 4);
                            pos += 4;
                            sb.append((char) Integer.parseInt(hex, 16));
                        }
                        default -> throw new IllegalArgumentException("invalid escape '\\" + esc + "'");
                    }
                } else {
                    sb.append(c);
                }
            }
            return sb.toString();
        }

        private JsonValue parseNumber() {
            int start = pos;
            if (peek() == '-') {
                pos++;
            }
            boolean isDouble = false;
            while (pos < s.length() && (Character.isDigit(s.charAt(pos)) || "+-.eE".indexOf(s.charAt(pos)) >= 0)) {
                char c = s.charAt(pos);
                if (c == '.' || c == 'e' || c == 'E') {
                    isDouble = true;
                }
                pos++;
            }
            String text = s.substring(start, pos);
            if (isDouble) {
                return new JsonValue(Double.parseDouble(text));
            }
            return new JsonValue(Long.parseLong(text));
        }
    }
}

File: sdks/local-rp/kotlin/src/test/kotlin/community/catalyst/linkkeys/localrp/kt/testutil/MiniJson.kt
package community.catalyst.linkkeys.localrp.kt.testutil

/**
 * A minimal, dependency-free JSON parser for reading the conformance vector
 * fixtures in tests -- ported from the Java sibling SDK's
 * `testutil.MiniJson` (test-scope only; this SDK's runtime dependency is
 * just the sibling Java SDK project, see README's "Architecture decision").
 *
 * Only what the JSON vector files under `sdks/local-rp/conformance` need:
 * objects, arrays, strings (with standard escapes), numbers, booleans, and
 * null.
 */
object MiniJson {
    fun parse(text: String): JsonValue {
        val p = Parser(text)
        p.skipWhitespace()
        val v = p.parseValue()
        p.skipWhitespace()
        require(p.atEnd()) { "trailing content after JSON value at position ${p.pos}" }
        return v
    }

    /** A parsed JSON value: object, array, string, number, boolean, or null. */
    class JsonValue internal constructor(private val raw: Any?) {
        fun isNull(): Boolean = raw == null

        fun asString(): String = raw as? String ?: error("not a string: $raw")

        fun asBoolean(): Boolean = raw as? Boolean ?: error("not a boolean: $raw")

        fun asLong(): Long = when (raw) {
            is Long -> raw
            is Double -> raw.toLong()
            else -> error("not a number: $raw")
        }

        @Suppress("UNCHECKED_CAST")
        fun asArray(): List<JsonValue> = raw as? List<JsonValue> ?: error("not an array: $raw")

        @Suppress("UNCHECKED_CAST")
        fun asObject(): Map<String, JsonValue> = raw as? Map<String, JsonValue> ?: error("not an object: $raw")

        fun get(key: String): JsonValue = asObject()[key] ?: error("missing key '$key' in $raw")

        fun getOrNull(key: String): JsonValue? = asObject()[key]

        /** `get(key).asString()`, or `null` if the key is absent or its value is JSON null. */
        fun getStringOrNull(key: String): String? = getOrNull(key)?.let { if (it.isNull()) null else it.asString() }
    }

    private class Parser(val s: String) {
        var pos = 0

        fun atEnd(): Boolean = pos >= s.length

        fun skipWhitespace() {
            while (pos < s.length && s[pos].isWhitespace()) pos++
        }

        private fun peek(): Char {
            require(pos < s.length) { "unexpected end of JSON input" }
            return s[pos]
        }

        private fun expect(c: Char) {
            require(peek() == c) { "expected '$c' at position $pos, got '${peek()}'" }
            pos++
        }

        fun parseValue(): JsonValue {
            skipWhitespace()
            return when (peek()) {
                '{' -> parseObject()
                '[' -> parseArray()
                '"' -> JsonValue(parseString())
                't' -> {
                    expectLiteral("true")
                    JsonValue(true)
                }
                'f' -> {
                    expectLiteral("false")
                    JsonValue(false)
                }
                'n' -> {
                    expectLiteral("null")
                    JsonValue(null)
                }
                else -> parseNumber()
            }
        }

        private fun expectLiteral(literal: String) {
            require(s.regionMatches(pos, literal, 0, literal.length)) { "expected '$literal' at position $pos" }
            pos += literal.length
        }

        private fun parseObject(): JsonValue {
            expect('{')
            val map = LinkedHashMap<String, JsonValue>()
            skipWhitespace()
            if (peek() == '}') {
                pos++
                return JsonValue(map)
            }
            while (true) {
                skipWhitespace()
                val key = parseString()
                skipWhitespace()
                expect(':')
                val value = parseValue()
                map[key] = value
                skipWhitespace()
                when (peek()) {
                    ',' -> pos++
                    '}' -> {
                        pos++
                        break
                    }
                    else -> error("expected ',' or '}' at position $pos")
                }
            }
            return JsonValue(map)
        }

        private fun parseArray(): JsonValue {
            expect('[')
            val list = ArrayList<JsonValue>()
            skipWhitespace()
            if (peek() == ']') {
                pos++
                return JsonValue(list)
            }
            while (true) {
                list.add(parseValue())
                skipWhitespace()
                when (peek()) {
                    ',' -> pos++
                    ']' -> {
                        pos++
                        break
                    }
                    else -> error("expected ',' or ']' at position $pos")
                }
            }
            return JsonValue(list)
        }

        private fun parseString(): String {
            expect('"')
            val sb = StringBuilder()
            while (true) {
                val c = s[pos++]
                if (c == '"') break
                if (c == '\\') {
                    when (val esc = s[pos++]) {
                        '"' -> sb.append('"')
                        '\\' -> sb.append('\\')
                        '/' -> sb.append('/')
                        'b' -> sb.append('\b')
                        'f' -> sb.append('\u000C')
                        'n' -> sb.append('\n')
                        'r' -> sb.append('\r')
                        't' -> sb.append('\t')
                        'u' -> {
                            val hex = s.substring(pos, pos + 4)
                            pos += 4
                            sb.append(hex.toInt(16).toChar())
                        }
                        else -> error("invalid escape '\\$esc'")
                    }
                } else {
                    sb.append(c)
                }
            }
            return sb.toString()
        }

        private fun parseNumber(): JsonValue {
            val start = pos
            if (peek() == '-') pos++
            var isDouble = false
            while (pos < s.length && (s[pos].isDigit() || "+-.eE".indexOf(s[pos]) >= 0)) {
                if (s[pos] == '.' || s[pos] == 'e' || s[pos] == 'E') isDouble = true
                pos++
            }
            val text = s.substring(start, pos)
            return if (isDouble) JsonValue(text.toDouble()) else JsonValue(text.toLong())
        }
    }
}
